### ISSUE TITLE
Better Typing for the fetchMore property

### ISSUE THREAD

**Issue by KyleMcNutt** [apollographql/apollo-feature-requests#165]
Relevant code line:
https://github.com/apollographql/react-apollo/blob/master/packages/common/src/types/types.ts#L62

TLDR: Currently the type for the `fetchMore` property is quite confusing to read, and possibly too complex? Could it be simplified and refactored so more of it is exported?

Much longer version:
I would love it if it could be extracted into more smaller types that could be exported separately so that they could be reused for testing purposes.

For example let's say I have code that looks like the following:

```
function ComponentThatGetsData() {
  const { data, fetchMore } = useGetQueryWithPaging({...});

  const onLoadMore = () => {
    // some logic...
    fetchMore({
      variables: {...},
      updateQuery: (previousData, { fetchMoreResult }) => {
        return {
          ...previousData,
          ...fetchMoreResult
        };
      }
    });
  };

  return <ComponentThatDoesPaging items={data} onLoadMoreItems={onLoadMore} />
}
```

Then if I want to test this Component, I might write out a mock for my hooks that looks like this:

```
jest
   .spyOn(moduleWithQueryHook, 'useGetQueryWithPaging')
    .mockImplementation(() => {
      return {
        ...boilerPlateApolloProps,
        data: {
          items: {
            list: [....],
            pageInfo: {
              skipToken: '...'
            }
          }
        },
        variables: {
          top: 10
        },
        // ~~~ Error here: config implicitly has the any type ~~~
        fetchMore: (config) => {
          const newPageOfItems: IGetItemsQuery = {...};
          const newResults = config.updateQuery(previousData, {
            fetchMoreResult: newPageOfItems,
            variables: {
              top: 10
            }
          });
          return new Promise<ApolloQueryResult<IGetItemsQuery >>(resolve => {
            resolve({
              ...boilerPlateApolloProps,
              data: newResults
            });
          });         
        },
        loading: false
      };
    });
```

Because I have the `noImplicitAny` flag turned on, on the TypeScript compiler, the `fetchMore` mock doesn't work because I haven't typed `config` to be anything.

So I look into the Apollo type to try and type it correctly, and hope to find a reusable type for `fetchMore` that I could drop in here. And I find this:

```
export type ObservableQueryFields<TData, TVariables> = Pick<
  ObservableQuery<TData, TVariables>,
  | 'startPolling'
  | 'stopPolling'
  | 'subscribeToMore'
  | 'updateQuery'
  | 'refetch'
  | 'variables'
> & {
  fetchMore: (<K extends keyof TVariables>(
    fetchMoreOptions: FetchMoreQueryOptions<TVariables, K> &
      FetchMoreOptions<TData, TVariables>
  ) => Promise<ApolloQueryResult<TData>>) &
    (<TData2, TVariables2, K extends keyof TVariables2>(
      fetchMoreOptions: { query?: DocumentNode } & FetchMoreQueryOptions<
        TVariables2,
        K
      > &
        FetchMoreOptions<TData2, TVariables2>
    ) => Promise<ApolloQueryResult<TData2>>);
};
```

So, `fetchMore` is only defined within the `ObservableQueryFields` type, and therefore, it isn't exported, so I can't reuse this within this test.

So I try to extract it myself, and put it into a test-helper, where I could then use it for any test where I need to test paging. And that ends up looking something like this:

```
type FetchMoreFunction<TData, TVariables> = (<K extends keyof TVariables>(
  fetchMoreOptions: FetchMoreQueryOptions<TVariables, K> &
    FetchMoreOptions<TData, TVariables>
) => Promise<ApolloQueryResult<TData>>) &
  (<TData2, TVariables2, K extends keyof TVariables2>(
    fetchMoreOptions: { query?: DocumentNode } & FetchMoreQueryOptions<
      TVariables2,
      K
    > &
      FetchMoreOptions<TData2, TVariables2>
  ) => Promise<ApolloQueryResult<TData2>>);
```

Which I initially thought would work as long as I take the `fetchMore` function that I had written inline to my mock above, and extract it above so it would look like this:

```
// ~~~ Error here: config implicitly has the any type ~~~
const fetchMore: FetchMoreFunction<IGetItemsQuery, IGetItemsQueryVariables> = (config) => {
  const newPageOfItems: IGetItemsQuery = {...};
  const newResults = config.updateQuery(previousData, {
    fetchMoreResult: newPageOfItems,
    variables: {
      top: 10
    }
  });
  return new Promise<ApolloQueryResult<IGetItemsQuery >>(resolve => {
    resolve({
      ...boilerPlateApolloProps,
      data: newResults
    });
  });         
};

jest
   .spyOn(moduleWithQueryHook, 'useGetQueryWithPaging')
    .mockImplementation(() => {
      return {
        ... // same thing from above...
        fetchMore,
        loading: false
      };
    });
```

But, the issue here is that the `config` type still can't be inferred by TypeScript, so I still have the implicit any problem.

So then I refactored my copy and pasted types again, so it looked like this now:

```
export type FetchMoreAndFetchMoreQueryOptions<
  TData,
  TVariables,
  K extends keyof TVariables
> = FetchMoreQueryOptions<TVariables, K> & FetchMoreOptions<TData, TVariables>;

type FetchMoreFunction<TData, TVariables> = (<K extends keyof TVariables>(
  fetchMoreOptions: FetchMoreAndFetchMoreQueryOptions<TData, TVariables, K>
) => Promise<ApolloQueryResult<TData>>) &
  (<TData2, TVariables2, K extends keyof TVariables2>(
    fetchMoreOptions: { query?: DocumentNode } & FetchMoreQueryOptions<
      TVariables2,
      K
    > &
      FetchMoreOptions<TData2, TVariables2>
  ) => Promise<ApolloQueryResult<TData2>>);
```

Which, I thought would be the end of it but, when I write my mock like this, I also get an error:

```
// ~~~ Error: Type 'IGetItemsQuery' is not assignable to type 'TData2'.
//                  'IGetItemsQuery' is assignable to the constraint of type 'TData2', but 'TData2'         //                  could be instantiated with a different subtype of constraint '{}' ~~~
const fetchMore: FetchMoreFunction<IGetItemsQuery, IGetItemsQueryVariables> = (config: FetchMoreAndFetchMoreQueryOptions<IGetItemsQuery, IGetItemsQueryVariables, 'other' | 'variables' | 'besides' | 'top'>) => {
  const newPageOfItems: IGetItemsQuery = {...};
  const newResults = config.updateQuery(previousData, {
    fetchMoreResult: newPageOfItems,
    variables: {
      top: 10
    }
  });
  return new Promise<ApolloQueryResult<IGetItemsQuery >>(resolve => {
    resolve({
      ...boilerPlateApolloProps,
      data: newResults
    });
  });         
};

jest
   .spyOn(moduleWithQueryHook, 'useGetQueryWithPaging')
    .mockImplementation(() => {
      return {
        ... // same thing from above...
        fetchMore,
        loading: false
      };
    });
```

And this is really where I am lost, what is the difference between TData2 and TData in this context? I know that my paginated data is always going return a consistent Query Result type, so why is it possible that it can be this other type (TData2)? Is there a way to constrain this type so it knows I only plan to return TData? Can we remove TData2 totally?

So from a new feature perspective, I'm hoping for two things:

1. Can we export the `fetchMore` type and `fetchMoreConfiguration` type so it could be reused in test code like this?
2. Can we simplify the `fetchMoreConfiguration` type, or allow for it to be constrained somehow, so that we don't have to deal with `TData2`?

**Comment by fbartho** [comment]
My team has been avoiding the `fetchMore`/Pagination case in part because of this. (Most of our APIs don't currently have pagination). Solving this would be nice.

**Comment by KyleMcNutt** [comment]
Note: As a workaround... I was able to use these types based on the type from Apollo:

```
import {
  ApolloQueryResult,
  FetchMoreOptions,
  FetchMoreQueryOptions
} from 'apollo-client';
import { DocumentNode } from 'apollo-link';

export type ApolloFetchMoreType<TData, TVariables> = (<
  K extends keyof TVariables
>(
  fetchMoreOptions: FetchMoreQueryOptions<TVariables, K> &
    FetchMoreOptions<TData, TVariables>
) => Promise<ApolloQueryResult<TData>>) &
  // Note I do not use TData2 or TVariables2
  (<K extends keyof TVariables>(
    fetchMoreOptions: {
      query?: DocumentNode;
    } & FetchMoreQueryOptions<TVariables, K> &
      FetchMoreOptions<TData, TVariables>
  ) => Promise<ApolloQueryResult<TData>>);

// This is a simpler type, that what is technically correct, but avoids the need for the "K" type
// from above, which uses a syntax only available when typing functions.
export type ApolloFetchMoreOptions<TData, TVariables> = {
  updateQuery: (
    previousData: TData,
    newData: {
      fetchMoreResult: TData;
      variables: TVariables;
    }
  ) => TData;
};
```
I use them in my test like this:
```
const fetchMore: ApolloFetchMoreType<IGetItemsQuery, IGetItemsQueryVariables> = (config: ApolloFetchMoreOptions<IGetItemsQuery, IGetItemsQueryVariables>) => {
  const newPageOfItems: IGetItemsQuery = {...};
  const newResults = config.updateQuery(previousData, {
    fetchMoreResult: newPageOfItems,
    variables: {
      top: 10
    }
  });
  return new Promise<ApolloQueryResult<IGetItemsQuery>>(resolve => {
    resolve({
      ...boilerPlateApolloProps,
      data: newResults
    });
  });         
};

jest
   .spyOn(moduleWithQueryHook, 'useGetQueryWithPaging')
    .mockImplementation(() => {
      return {
        ...boilerPlateApolloProps,
        data: {
          items: {
            list: [....],
            pageInfo: {
              skipToken: '...'
            }
          }
        },
        variables: {
          top: 10
        },
        fetchMore,
        loading: false
      };
    });
```
This still isn't a great solution, but it is a workaround.

Ideally, it would be great to not have to copy and paste types from Apollo, and also to not (sort of) invent new types, that don't include `K` or `TData2`/`TVariables2`.

**Comment by lhguerra** [comment]
This would be really nice. I'm currently killing a few baby pandas and typing it this way:
```
type MyNiceType = {
  loading: boolean
  posts?: Post[]
  fetchMore: (props: any) => void
}
```

I'm not going further than that for now because I don't really understand what I should guarantee with this type.

**Comment by younes0** [comment]
@jerelmiller the apollo team actually created a FetchMoreFunction type with the release of `useSuspenseQuery`

```tsx
import { FetchMoreFunction } from "@apollo/client/react/hooks/useSuspenseQuery";

type MyCustomFetchMore = FetchMoreFunction<MyQuery, MyQueryVariables>;
```

It's a bit odd to import this from `useSuspenseQuery` file since it's compatible with `useQuery`

**Comment by jerelmiller** [comment]
@younes0 keen eye you've got for that type 😄. I kept it "private" (i.e. not exported in the main export) is because I'd like to see if I can refactor this type a bit more on the underlying `ObservableQuery` type to avoid the need for this extra function type. This type works well for the underlying `useSuspenseQuery` hook, but I'd like to eliminate the need for it if possible. Feel free to use it for now, but I'd consider it "unstable" as it could be moved/removed in a future version.

**Comment by younes0** [comment]
@jerelmiller got it, thanks for your reply
